fido/Crypto/Fido2/PublicKey.hs

```diff
@@ -0,0 +1,279 @@
+  xKey <- CBOR.decodeIntCanonical
+  when (xKey /= -2) $ fail "Expected required `x`"
+  x <- os2ip <$> CBOR.decodeBytesCanonical
+  yKey <- CBOR.decodeIntCanonical
+  when (yKey /= -3) $ fail "Expected required `x`"
+  tokenType <- CBOR.peekTokenType
+  y <- case tokenType of
+    -- TODO(arianvp): Implement compressed curve. Waiting for
+    -- https://github.com/haskell-crypto/cryptonite/issues/302
+    CBOR.TypeBool -> fail "Compressed format not supported _yet_ See Issue number X"
+    -- direct coordinate
+    CBOR.TypeBytes -> os2ip <$> CBOR.decodeBytesCanonical
+    _ -> fail "Unexpected token type"
+  let point = ECC.Point x y
+  when (not (ECC.isPointValid curve point)) $ fail "point not on curve"
+  pure $ ECDSAKey hash (ECDSA.PublicKey curve (ECC.Point x y))
+
+encodePublicKey :: PublicKey -> Encoding
+encodePublicKey (ECDSAPublicKey (ECDSAKey alg (ECDSA.PublicKey curve point))) =
+  CBOR.encodeMapLen 5
+    <> CBOR.encodeInt 1
+    <> CBOR.encodeInt 2
+    <> CBOR.encodeInt 3
+    <> encodeCOSEAlgorithmIdentifier (ECDSAIdentifier alg)
+    <> CBOR.encodeInt (-1)
+    <> CBOR.encodeInt
+      ( case curve of
+          curve | curve == ECC.getCurveByName ECC.SEC_p256r1 -> 1
+          curve | curve == ECC.getCurveByName ECC.SEC_p384r1 -> 2
+          curve | curve == ECC.getCurveByName ECC.SEC_p521r1 -> 3
+          _ | otherwise -> error "never happens"
+      )
+    <> ( case point of
+           ECC.Point x y ->
+             CBOR.encodeInt (-2)
+               <> CBOR.encodeBytes (i2osp x)
+               <> CBOR.encodeInt (-3)
+               <> CBOR.encodeBytes (i2osp y)
+           _ -> error "never happens"
+       )
+encodePublicKey (EdDSAPublicKey key) =
+  CBOR.encodeMapLen 4
+    <> CBOR.encodeInt 1
+    <> CBOR.encodeInt 1
+    <> CBOR.encodeInt 3
+    <> encodeCOSEAlgorithmIdentifier EdDSA
+    <> CBOR.encodeInt (-1)
+    <> case key of
+      Ed25519 key ->
+        CBOR.encodeInt 6
+          <> CBOR.encodeInt (-2)
+          <> CBOR.encodeBytes (ByteArray.convert key)
+      Ed448 key ->
+        CBOR.encodeInt 7
+          <> CBOR.encodeInt (-2)
+          <> CBOR.encodeBytes (ByteArray.convert key)
+
+encodeCOSEAlgorithmIdentifier :: COSEAlgorithmIdentifier -> Encoding
+encodeCOSEAlgorithmIdentifier x = CBOR.encodeInt $ case x of
+  ECDSAIdentifier ES256 -> (-7)
+  ECDSAIdentifier ES384 -> (-35)
+  ECDSAIdentifier ES512 -> (-36)
+  EdDSA -> (-8)
+
+-- | Decodes a signature for a specific public key's type
+-- Signatures are a bit weird in Webauthn.  For ES256 and RS256 they're ASN.1
+-- and for EdDSA they're COSE
+decodeECDSASignature :: ByteString -> Maybe ECDSA.Signature
+decodeECDSASignature sigbs =
+  case ASN1.decodeASN1' ASN1.BER sigbs of
+    Left _ -> Nothing
+    Right [ASN1.Start ASN1.Sequence, ASN1.IntVal r, ASN1.IntVal s, ASN1.End ASN1.Sequence] ->
+      Just (ECDSA.Signature r s)
+    Right _ -> Nothing
+
+verify :: PublicKey -> ByteString -> ByteString -> Bool
+verify key msg sig =
```

Comment:
One downside of a `verify` that returns a bool is that you can forget to call it. One way to mitigate this (until we get Linear Haskell :smile:) is to fuse deserialization of the message with verification, something like

```haskell
verifyAndDecode :: PublicKey -> Message ByteString -> Signature -> Maybe DecodedMessage
```
is there a sensible decode operation for the message to do this with?

A different option would be to return some kind of validity token of which the constructor is not exposed, and to have remaining functions take such a token as input.

Reply:
No not really.  `msg =  authenticatorDataBytes || sha256(clientDataJSONBytes)`
so basically involves concatenating two unparsed values, instead of doing calculations on our
parsed datatypes.  There's not much sensible stuff to parse out of it; as at that point we already have the parsed `authenticatorData` and `clientData` anyway

https://github.com/arianvp/haskell-fido2/blob/master/fido/Crypto/Fido2/Assertion.hs#L98

Reply:
Hmm yeh we could still do this...  i'll add this as a todo for later as IDK what is the most ergonomic.

Both attestation and assertion use the same signature format so we have that going for us I think